### ISSUE TITLE
Make `MetricsLambda` update it's underlying metrics

### ISSUE THREAD

**Issue by Moh-Yakoub** [pytorch/ignite#1921]
## 🚀 Feature

As title.

The current state is that we cannot use `MetricsLambda` as a standard metric because it doesn't update/reset it's underlying metrics.
Adding this will expand the scope and usability of `MetricsLambda`

I've already an idea how to implement this based on our discussion https://github.com/pytorch/ignite/pull/1887#discussion_r610902667

So @vfdev-5 please assign the issue to me.

**Comment by sdesrozis** [comment]
I'm not sure that LambdaMetric should update its related metrics. If so, what happen if the metrics are attached to an engine ? I remember past discussions on that point and we stated that the update of metrics must be on user side.

**Comment by vfdev-5** [comment]
@sdesrozis yes that's true. However, it means that our metrics arithmetics does not work with reset/update/compute API : 
```python
import torch
import ignite
print(ignite.__version__)

from ignite.metrics import Accuracy

err = (1.0 - Accuracy()) * 100.0

y_preds = torch.rand(4, 10)
y = torch.randint(0, 10, size=(4, ))

err.update((y_preds, y))
err.compute()
> NotComputableError
```

An idea could be to make `MetricsLambda` aware of the fact that it is attached or not and dispatch `update` according to that. What do you think ?

**Comment by Moh-Yakoub** [comment]
> @sdesrozis yes that's true. However, it means that our metrics arithmetics does not work with reset/update/compute API :
> 
> ```python
> import torch
> import ignite
> print(ignite.__version__)
> 
> from ignite.metrics import Accuracy
> 
> err = (1.0 - Accuracy()) * 100.0
> 
> y_preds = torch.rand(4, 10)
> y = torch.randint(0, 10, size=(4, ))
> 
> err.update((y_preds, y))
> err.compute()
> > NotComputableError
> ```
> 
> An idea could be to make `MetricsLambda` aware of the fact that it is attached or not and dispatch `update` according to that. What do you think ?

That can be an option. Another idea can be that we have a new subclass called `ComputableMetricsLambda` or something like this and it won't allow engines to be attached to it.

**Comment by vfdev-5** [comment]
>  Another idea can be that we have a new subclass called ComputableMetricsLambda or something like this and it won't allow engines to be attached to it.

This can be ambiguous, I think. For example, `err = (1.0 - Accuracy()) * 100.0` would be `MetricsLambda` or `ComputableMetricsLambda` ? User can create a new metric like that and could use it in both scenarios (separately)...

**Comment by sdesrozis** [comment]
I understand but I'm not sure...

Metrics `LambdaMetric` should check if its metrics would be already attached or not every update calls. Moreover, it means metrics would have to handle a state concerning attachment or not, and detachment.

A simple alternative would be to pass metrics to the report and not construct and attached internally... 

IMO a report is not a metric. It uses metrics.

**Comment by vfdev-5** [comment]
> A simple alternative would be to pass metrics to the report and not construct and attached internally...

For me, the main point of the issue here is to solve this issue : https://github.com/pytorch/ignite/issues/1921#issuecomment-817144804

**Comment by sdesrozis** [comment]
> An idea could be to make `MetricsLambda` aware of the fact that it is attached or not and dispatch `update` according to that. What do you think ?

I need to write codes to go further on this

**Comment by sdesrozis** [comment]
> For me, the main point of the issue here is to solve this issue : https://github.com/pytorch/ignite/issues/1921#issuecomment-817144804
> 
> ```python
> 
> err = (1.0 - Accuracy()) * 100.0
> 
> ```
 
It's a good use case. 

Here, we have one `LambdaMetric` per operator (minus and mult) defined by the tree op. Call `err.update(...)` means recursive checks on the tree where the leaf is `Accuracy()`. This metric is not attached so the `update(...)` should automatically called. 

It looks simple but error prone
```
acc = Accuracy()
err = (1.0 - acc) * 100.0

y_preds = torch.rand(4, 10)
y = torch.randint(0, 10, size=(4, ))

acc.update((y_preds, y))
err.update((y_preds, y))
err.compute()
```
